File: 509-fibonacci-number/509-fibonacci-number.py
class Solution:
    def fib(self, n: int) -> int:
        if n <= 1:
            return n
        dp = [-1 for _ in range(n+1)]
        dp[0] = 0
        dp[1] = 1
        def recFibo(k: int) -> int:
            if dp[k] != -1:
                return dp[k]
            res = recFibo(k-1) + recFibo(k-2)
            dp[k] = res
            return res
        return recFibo(n)

File: 509-fibonacci-number/509-fibonacci-number.rs
impl Solution {
    pub fn fib(n: i32) -> i32 {
        let five = 5 as f32;
        let sqrt5 = five.sqrt();
        let phi = (1 as f32 + sqrt5) / (2 as f32);
        let res = (f32::powf(phi, n as f32) / sqrt5).round();
        return res as i32;
    }
}